core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -106,7 +107,66 @@
         /**
          * Unassigned as a result of a failed primary while the replica was initializing.
          */
-        PRIMARY_FAILED;
+        PRIMARY_FAILED
+    }
+
+    /**
+     * Captures the status of an unsuccessful allocation attempt for the shard,
+     * causing it to remain in the unassigned state.
+     *
+     * Note, ordering of the enum is important, make sure to add new values
+     * at the end and handle version serialization properly.
+     */
+    public enum AllocationAttempt implements Writeable {
+        /**
+         * The shard was denied allocation to a node because the allocation deciders all returned a NO decision
+         */
+        DENIED_BY_ALLOCATION_DECIDERS,
```

Comment:
can we add explicit ids for these, or have a test that makes sure the ids don't change? we're one minor mistake away from breaking the wire protocol